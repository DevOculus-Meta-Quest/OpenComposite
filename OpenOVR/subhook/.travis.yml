language: cpp

compiler:
  - gcc
  - clang

env:
  global:
    - CTEST_OUTPUT_ON_FAILURE=ON
  matrix:
    - BITS=32
    - BITS=64

install:
  - sudo apt-get update -qq
  - sudo apt-get install -y cmake gcc-multilib g++-multilib

before_script:
  - cmake . -DCMAKE_C_COMPILER=${CC}
            -DCMAKE_CXX_COMPILER=${CXX}
            -DCMAKE_C_FLAGS=-m$BITS
            -DCMAKE_CXX_FLAGS=-m$BITS

script:
  - make
  - make test
