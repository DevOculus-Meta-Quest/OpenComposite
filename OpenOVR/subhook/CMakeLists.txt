cmake_minimum_required(VERSION 2.8.6)
cmake_policy(SET CMP0042 NEW)

project(subhook)

set(SUBHOOK_VERSION_MAJOR 0)
set(SUBHOOK_VERSION_MINOR 4)
set(SUBHOOK_VERSION_PATCH 0)

set(SUBHOOK_VERSION ${SUBHOOK_VERSION_MAJOR})
set(SUBHOOK_VERSION ${SUBHOOK_VERSION}.${SUBHOOK_VERSION_MINOR})
set(SUBHOOK_VERSION ${SUBHOOK_VERSION}.${SUBHOOK_VERSION_PATCH})

set(SUBHOOK_HEADERS subhook.h)
set(SUBHOOK_SOURCES subhook.c subhook_private.h)

add_definitions(-DSUBHOOK_IMPLEMENTATION)

if(SUBHOOK_STATIC)
  add_library(subhook STATIC ${SUBHOOK_HEADERS} ${SUBHOOK_SOURCES})
else()
  add_library(subhook SHARED ${SUBHOOK_HEADERS} ${SUBHOOK_SOURCES})
endif()

include_directories(${CMAKE_CURRENT_SOURCE_DIR})
set_property(DIRECTORY ${CMAKE_SOURCE_DIR}
             APPEND PROPERTY INCLUDE_DIRECTORIES ${CMAKE_CURRENT_SOURCE_DIR})

if(SUBHOOK_STATIC)
  add_definitions(-DSUBHOOK_STATIC)
  set_property(DIRECTORY ${CMAKE_SOURCE_DIR}
               APPEND PROPERTY COMPILE_DEFINITIONS SUBHOOK_STATIC)
endif()

set_property(GLOBAL PROPERTY USE_FOLDERS ON)

if(NOT SUBHOOK_NO_INSTALL)
  install(TARGETS subhook LIBRARY DESTINATION lib
                          ARCHIVE DESTINATION lib
                          RUNTIME DESTINATION bin)
  install(FILES ${SUBHOOK_HEADERS} DESTINATION include)
endif()

set(CPACK_PACKAGE_NAME ${PROJECT_NAME})
set(CPACK_PACKAGE_VERSION_MAJOR ${SUBHOOK_VERSION_MAJOR})
set(CPACK_PACKAGE_VERSION_MINOR ${SUBHOOK_VERSION_MINOR})
set(CPACK_PACKAGE_VERSION_PATCH ${SUBHOOK_VERSION_PATCH})

include(CPack)
include(CTest)

if(BUILD_TESTING AND NOT SUBHOOK_NO_TESTS)
  enable_testing()
  add_subdirectory(test)
endif()
