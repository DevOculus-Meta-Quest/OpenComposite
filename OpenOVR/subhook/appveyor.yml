version: '{build}'

platform:
  - x86
  - x64

configuration:
  - Release

environment:
  CTEST_OUTPUT_ON_FAILURE: ON

before_build:
  - cmake .

build_script:
  - cmake --build . --config %CONFIGURATION%
