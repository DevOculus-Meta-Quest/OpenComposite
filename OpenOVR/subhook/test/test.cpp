#include <cstdio>
#include <subhook.h>

typedef void (*foo_func_t)();

foo_func_t foo_tr = 0;

void foo() {
  std::printf("foo() called\n");
}

void foo_hooked() {
  std::printf("foo_hooked() called\n");
}

void foo_hooked_tr() {
  std::printf("foo_hooked_tr() called\n");
  foo_tr();
}

int main() {
  std::printf("Testing initial install\n");

  subhook::Hook foo_hook;
  if (!foo_hook.Install((void *)foo,
                        (void *)foo_hooked,
                        subhook::HookOption64BitOffset)) {
    std::printf("Install failed\n");
    return 1;
  }
  foo();
  if (!foo_hook.Remove()) {
    std::printf("Remove failed\n");
    return 2;
  }
  foo();

  std::printf("Testing re-install\n");

  if (!foo_hook.Install()) {
    std::printf("Install failed\n");
    return 3;
  }
  foo();
  if (!foo_hook.Remove()) {
    std::printf("Remove failed\n");
    return 4;
  }
  foo();

  std::printf("Testing trampoline\n");

  subhook::Hook foo_hook_tr((void *)foo,
                            (void *)foo_hooked_tr,
                            subhook::HookOption64BitOffset);
  if (!foo_hook_tr.Install()) {
    std::printf("Install failed\n");
    return 5;
  }
  foo_tr = (foo_func_t)foo_hook_tr.GetTrampoline();
  foo();
}
