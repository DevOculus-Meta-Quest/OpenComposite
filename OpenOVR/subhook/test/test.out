Testing initial install
foo_hooked\(\) called
foo\(\) called
Testing re-install
foo_hooked\(\) called
foo\(\) called
Testing trampoline
foo_hooked_tr\(\) called
foo\(\) called
